pyoxenmq overhaul

jagerman opened the pull request:
- Wrap much more of the OxenMQ
- Add extensive pydoc to everything
- Switch setup.py to pybind's built-in tools

This was https://github.com/oxen-io/oxen-mq/pull/61 but it proved better for deb packaging to keep it in a separate repo.

jagerman committed:
pyoxenmq overhaul

- Wrap much more of the OxenMQ
- Add extensive pydoc to everything
- Switch setup.py to pybind's built-in tools

majestrate commented on CMakeLists.txt:
```diff
@@ -45,11 +49,15 @@ if(SUBMODULE_CHECK)
-add_subdirectory(pyoxenmq)
+pybind11_add_module(oxenmq MODULE
+    src/bencode.cpp
+    src/bencode.cppoxenmq.cpp
```
love that filename, `bencode.cppoxenmq.cpp`

majestrate commented:
additionally a ci pipeline for drone would be useful

jagerman committed:
Remove unwanted headers

- pybind11/gil.h is only in recent pybind (but is implicitly included by
  the main pybind11.h header)
- detail/common shouldn't be included

jagerman committed:
De-Frankenstein the build

jagerman commented:
Built and pushed debs from this branch to our apt repo.

Will figure out some (non-deb) pipelines tomorrow.

jagerman committed:
Formatting

jagerman committed:
Remove cmake build system and submodules

setup.py works fine, and we don't need two separate build systems.

Remove the submodules as well because we aren't using them and don't
need them; you should install pybind11 and oxenmq outside this project.

jagerman committed:
Various fixes for bugs revealed in testing

- make extract data parts always flatten to arbitrary depth.  We were
  only flattening once (and only sometimes), but arbitrary flattening is
  nicer (as long as we don't ultimately encouter anything that isn't
  str/bytes).
- fix Message.data()/.dataview() pre-loading nulls into the list.  The
  py::list constructor argument here is pre-fill, not a reservation.
- fix add_command: it was broken because when the wrapped callback gets
  invoked it tries casting the Message via copying, but that breaks
  (Message isn't copyable).  Fix it by adding a wrapper that does a
  referencing cast to a Python object.
- Add missing docs for add_command
- Doc typo fixes
- Remove OxenMQ default and loglevel constructors, and make the
  remaining constructor take everything by keyword arguments.  (Default
  constructor remains equivalent, and the log-level constructor now has
  to have the log level named).
- Fix OxenMQ keyword constructor: invoking it was failing because the
  annotation default on sn_lookup wasn't properly castable to the
  required types (it should be py::none() rather than nullptr).  Also
  fix the defaults for pubkey/privkey to be py::bytes() (the empty
  string worked fine, but using bytes makes it show up as bytes defaults
  in the generated signature doc string).
- Further document connect_inproc with a description of why you might
  want it.
- Fix send not actually including the message data parts
- Work around segfault in the reply callback: because we have lambdas
  with python object captures, we'd segfault in the proxy thread when it
  frees them after calling them because that freeing invokes python
  destructor but it doesn't hold the gil.  Fixed by making the lambda
  consume its own callbacks (which is fine because oxenmq will invoke
  the callback exactly once).
- Fix request_future invocation which raised an error on invocation:
  straight lambdas aren't castable to python objects, so stuff the
  lambda inside std::functions, which are.

jagerman committed:
Add simple pytest tests

jagerman committed:
Add drone build & tests

jagerman committed:
Hacking around for macos

Build local oxenmq

Install everything into a local prefix

jagerman committed:
Build with -Wextra -Werror and color

jagerman committed:
Fixes and address warnings

- timeout in the syncronous connect_remote wasn't being passed through
- callback and syncronous connect_remote had some deviation in args;
  unified them to both take keyword-only timeout and
  ephemeral_routing_id, and to take timeout as a timedelta with a
  default (rather than an optional in the callback-based one).
- removed unused "noopt".  I forget what it was for, but it didn't end
  up being used.
